app/src/mouse/input_processor_temp_layer.c

```diff
@@ -9,52 +9,182 @@
+        return ZMK_EV_EVENT_BUBBLE;
     }
+
+    const struct device *dev = DEVICE_DT_INST_GET(0);
+    struct temp_layer_data *data = (struct temp_layer_data *)dev->data;
+    data->state.last_tapped_timestamp = ev->timestamp;
+
+    return ZMK_EV_EVENT_BUBBLE;
 }
 
-static int tl_handle_event(const struct device *dev, struct input_event *event, uint32_t param1,
-                           uint32_t param2, struct zmk_input_processor_state *state) {
-    if (param1 >= ZMK_KEYMAP_LAYERS_LEN) {
-        LOG_ERR("Passed an invalid layer id %d", param1);
+/* Driver Implementation */
+static int temp_layer_handle_event(const struct device *dev, struct input_event *event,
+                                   uint32_t param1, uint32_t param2,
+                                   struct zmk_input_processor_state *state) {
+    if (param1 >= MAX_LAYERS) {
+        LOG_ERR("Invalid layer index: %d", param1);
         return -EINVAL;
     }
 
-    if (!zmk_keymap_layer_active(param1)) {
-        zmk_keymap_layer_activate(param1);
+    struct temp_layer_data *data = (struct temp_layer_data *)dev->data;
+    const struct temp_layer_config *cfg = dev->config;
+
+    data->state.toggle_layer = param1;
+
+    if (!data->state.is_active &&
+        !should_quick_tap(cfg, data->state.last_tapped_timestamp, k_uptime_get())) {
+        update_layer_state(&data->state, true);
+    }
+
+    if (param2 > 0) {
+        k_work_reschedule(&layer_disable_works[param1], K_MSEC(param2));
     }
 
-    k_work_reschedule(&layer_disable_works[param1], K_MSEC(param2));
     return 0;
 }
 
-static struct zmk_input_processor_driver_api tl_driver_api = {
-    .handle_event = tl_handle_event,
-};
+static int temp_layer_init(const struct device *dev) {
+    struct temp_layer_data *data = dev->data;
+    data->dev = dev;
+    data->state = (struct temp_layer_state){0};
 
-static int tl_init(const struct device *dev) {
-    for (int i = 0; i < ZMK_KEYMAP_LAYERS_LEN; i++) {
-        k_work_init_delayable(&layer_disable_works[i], layer_disable_cb);
+    for (int i = 0; i < MAX_LAYERS; i++) {
+        k_work_init_delayable(&layer_disable_works[i], layer_disable_callback);
     }
 
+    LOG_INF("Auto layer processor initialized");
     return 0;
 }
 
-#define TL_INST(n)                                                                                 \
-    DEVICE_DT_INST_DEFINE(n, &tl_init, NULL, NULL, NULL, POST_KERNEL,                              \
-                          CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, &tl_driver_api);
+/* Driver API */
+static const struct zmk_input_processor_driver_api temp_layer_driver_api = {
+    .handle_event = temp_layer_handle_event,
+};
+
+/* Event Listeners */
+ZMK_LISTENER(processor_temp_layer, handle_position_state_changed);
+ZMK_SUBSCRIPTION(processor_temp_layer, zmk_position_state_changed);
+ZMK_LISTENER(processor_temp_layer_keycode, handle_keycode_state_changed);
+ZMK_SUBSCRIPTION(processor_temp_layer_keycode, zmk_keycode_state_changed);
```

Comment:
This shouldn't be two listeners, it should be one listener with multiple subscriptions, and then the callback should handle determining which type of event this is.

I'm a *little* annoyed at having this not be conditional on this feature actually being needed/used, but don't see a way to avoid it.

Reply:
```
/* Event Listeners Conditions */
#define NEEDS_POSITION_HANDLERS(n) DT_INST_PROP_HAS_IDX(n, excluded_positions, 0)
#define NEEDS_KEYCODE_HANDLERS(n) (DT_INST_PROP_OR(n, require_prior_idle_ms, 0) > 0)

/* Event Handlers Registration */
#if DT_INST_FOREACH_STATUS_OKAY_VARGS(NEEDS_POSITION_HANDLERS, ||) || \
    DT_INST_FOREACH_STATUS_OKAY_VARGS(NEEDS_KEYCODE_HANDLERS, ||)
    ZMK_LISTENER(processor_temp_layer, handle_state_changed_dispatcher);
#endif

/* Individual Subscriptions */
#if DT_INST_FOREACH_STATUS_OKAY_VARGS(NEEDS_POSITION_HANDLERS, ||)
    ZMK_SUBSCRIPTION(processor_temp_layer, zmk_position_state_changed);
#endif

#if DT_INST_FOREACH_STATUS_OKAY_VARGS(NEEDS_KEYCODE_HANDLERS, ||)
    ZMK_SUBSCRIPTION(processor_temp_layer, zmk_keycode_state_changed);
#endif
```

does something like this work

Reply:
Something like that should be possible, yeah.  You'll need to play a bit.

Reply:
looks like it works, snapshots from tests shows:
require-idle-ms & excluded_positions not set -> no events
require-idle-ms set -> dispatch + keycode event triggered and no position event
excluded_positions set -> dispatch + position event triggered and no keycode event